kernel/sleep/src/lib.rs

```diff
@@ -9,128 +9,219 @@
-        match delayed_tasklist.peek() {
-            Some(SleepingTaskNode { resume_time, .. }) => 
-                NEXT_DELAYED_TASK_UNBLOCK_TIME.store(*resume_time, Ordering::SeqCst),
-            None => NEXT_DELAYED_TASK_UNBLOCK_TIME.store(usize::MAX, Ordering::SeqCst),
-        }
+/// Awaken all sleepers who have hit their wake time.
+pub fn awaken_sleepers() {
+    let time = tick_count();
+
+    // Check the next wake time. If it's in the future we don't need to get the `SLEEPERS` lock.
+    let mut next_wake_time = NEXT_WAKE_TIME.load(Ordering::SeqCst);
+    if next_wake_time > time {
+        return;
+    } 
+
+    // Wake up all sleepers who've hit their wake time.
+    let mut sleepers = SLEEPERS.lock();
+    while next_wake_time <= time {
+        let Some(mut sleeper) = sleepers.pop() else {
+            return;
+        };
+
+        sleeper.awaken();
+
+        next_wake_time = match sleepers.peek() {
+            Some(next_sleeper) => next_sleeper.wake_time,
+            None => Ticks::MAX,
+        };
     }
-}
 
-/// Remove all tasks that have been delayed but are able to be unblocked now,
-/// the current tick count is provided by the system's interrupt tick count.
-pub fn unblock_sleeping_tasks() {
-    let ticks = TICK_COUNT.load(Ordering::SeqCst);
-    while ticks > NEXT_DELAYED_TASK_UNBLOCK_TIME.load(Ordering::SeqCst) {
-        remove_next_task_from_delayed_tasklist();
-    }
+    // Store the earliest sleeper's wake time for next time.
+    NEXT_WAKE_TIME.store(next_wake_time, Ordering::SeqCst);
 }
 
-/// Blocks the current task by putting it to sleep for `duration` ticks.
-pub fn sleep(duration: usize) {
-    let current_tick_count = TICK_COUNT.load(Ordering::SeqCst);
-    let resume_time = current_tick_count + duration;
 
-    let current_task = current_task().clone();
-    // Add the current task to the delayed tasklist and then block it.
-    add_to_delayed_tasklist(SleepingTaskNode{taskref: current_task.clone(), resume_time});
-    current_task.block();
-    scheduler::yield_now();
+
+/// Blocks asynchronously for `duration` ticks.
+pub async fn sleep(duration: Ticks) {
+    sleep_until(tick_count() + duration).await;
 }
 
-/// Blocks the current task by putting it to sleep until a specific tick count is reached,
-/// given by `resume_time`.
-pub fn sleep_until(resume_time: usize) {
-    let current_tick_count = TICK_COUNT.load(Ordering::SeqCst);
+/// Blocks asynchronously until a specific absolute time.
+/// If the time has already passed there is no delay.
+pub fn sleep_until(wake_time: Ticks) -> impl Future<Output = ()> {
+    let sleeper = Sleeper::new(wake_time);
 
-    if resume_time > current_tick_count {
-        sleep(resume_time - current_tick_count);
+    let mut sleepers = SLEEPERS.lock();
+    sleepers.push(sleeper.clone());
+
+    let next_wake_time = NEXT_WAKE_TIME.load(Ordering::SeqCst);
+    if sleeper.wake_time < next_wake_time {
+        NEXT_WAKE_TIME.store(sleeper.wake_time, Ordering::SeqCst);
```

Comment:
```suggestion
    let _ = NEXT_WAKE_TIME.fetch_update(Ordering::SeqCst, Ordering::SeqCst, |next_wake_time| {
        if sleeper.wake_time < next_wake_time {
            Some(sleeper.wake_time)
        } else {
            None
        }
    });
```
I'm pretty sure the original code has a race condition in it. If the task is preempted between the load and the store, `next_wake_time` might become outdated.

Reply:
Interrupts are disabled so there shouldn't be any preemption. I'll take a closer look at your suggestion when I get back from vacation. If it's equivalent and doesn't rely on preemption being disabled, all the better.

Reply:
I don't think interrupts are disabled in this function. Even if they are, another core could mutate `NEXT_WAKE_TIME` between the two operations.